More Adjoint Variable Diagnostics

timothyas opened the pull request:
## What changes does this PR introduce?
(Bug fix, feature, docs update, ...)

This adds to the previous adjoint diagnostics infrastructure in a couple of ways:
- Add all variables in `addummy_in_stepping`
- Add adjoint ptracer variables
- Rewrite of `addummy_in_stepping` to make it ... more compact and easier to add new control variables

## What is the current behaviour? 
(You can also link to an open issue here)


## What is the new behaviour 
(if this is a feature change)?


## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)

I hope not. I did a little bit of testing and it seems to be working

## Other information:

This is a work in progress, some other items would be:
- [x] Currently both `RS` and `RL` variables are dumped. I'm not sure how to handle `RS` variables with diagnostics bec other than copying them to a temporary variable then printing. This is already done if `dumpAdVarExch`=2, but perhaps this should always be done for `RS` variables. Actually, that seems straight forward...
- [x] add `dumpAdRec` as parameter to the `dump_adj` routines
- [x] Add add seaice and exf fields
- [x] merge the `ptracers_ad_dump` with the new `dump_adj*` routines
- [x] verification setup for variables in addummy_in_stepping 
- [x] verification setup for variables seaice and exf fields
- [x] verification setup for variables ptracers fields
- [x] dig into frequency=1., error
- [x] add documentation regarding adj etan

As mentioned in MITgcm/xmitgcm#247, using the diagnostics package is nice because xmitgcm can read these adjoint fields "out of the box" (i.e. without adding meta data to [xmitgcm/variables.py](https://github.com/MITgcm/xmitgcm/blob/master/xmitgcm/variables.py) ... which is good, because anyone using the adjoint needs all the help they can get!

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)

- Added all current adjoint variables in addummy_in_stepping, and ptracers to work with diagnostics package
- Streamlined addummy_in_stepping for mdsio output

timothyas commented:
Pinging @mjlosch since he was interested :)
I'll get to those to-do notes when I can.

mjlosch commented:
Hi @timothyas 
this looks great. There's actually a routine `DIAGNOSTICS_FILL_RS`, see recent commit 35cf6b5  , so we can probably tick that box. There are a couple of things that I noticed:
- ADJetan is written elsewhere (addummy_for_etan), can we add your machinery there as well?
- we should have the code tested in verification experiments as much as possible. I can see that there's something in global_ocean.cs32x15/input_ad, maybe it's worth it to add a second stream with 2D and RS/RL variables. I suggest: 'ADJetan ', 'ADJqnet ', 'ADJempmr', 'ADJtaux ','ADJtauy '. Anything else?
- for ` frequency(3) = 1., `, the model stalls with a Floating point exception: 
```* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_ARITHMETIC (code=EXC_I386_SSEEXTERR, subcode=0x1921)
    frame #0: 0x0000000100ee7642 mitgcmuv_ad`diagnostics_out_ at diagnostics_out.for:4109
   4106	C     a) find the time of the previous multiple of output freq:
   4107	        timeRec(1) = myTime-deltaTClock*0.5D0
   4108	        timeRec(1) = (timeRec(1)-phase(listId))/freq(listId)
-> 4109	        i = INT( timeRec(1) )
   4110	        IF ( timeRec(1).LT.0. ) THEN
   4111	          tmpLoc = FLOAT(i)
   4112	          IF ( timeRec(1).NE.tmpLoc ) i = i - 1
```
any idea why that's not allowed? `-1.` seems to work.
- We should definitely add seaice and exf fields. I can try to do that, assuming that it happens in seaice_ad_dump.F and exf_adjoint_snapshots_ad.F, right?

mjlosch commented:
@timothyas It looks like the global variable `dumpAdRecMn` is different for seaice (`dumpAdRecSi`) and exf (`dumpAdRecEx`). This means that we need to pass this as an argument to all dump_adj_*.F routines, right?

timothyas commented:
Thanks for taking a look @mjlosch! I updated the to-do list with a couple of your notes. Here are a couple of replies:

> ADJetan is written elsewhere (addummy_for_etan), can we add your machinery there as well?

This one is a bit difficult because the time step is "off by one" with the other variables. Here is what I wrote in #151 

>> The caveat here is that I could not get this to work with the diagnostics package, and this issue is a bit technical. I put the dummy routine in this particular location so that adEtaN gets the impact of adjoint variables from calc_r_star_ad or calc_surf_dr_ad, and before it is integrated through the adjoint model, where adEtaN is zero'd out as sensitivities are passed to other variables. However, at this particular spot adEtaN is a "half" time step off from the other variables, so it did not make sense to fill the diagnostic slots for ADJetan to be printed later via addummy_in_stepping because they would be inconsistent. I also could not think of a way to print individual diagnostic variables at different times. This could be fixed by placing dummy_in_stepping in a similar manner to do_state_vars_diags but for adjoint variables, but this seemed like overkill for a first step.

I remember thinking that the adjoint diagnostics would formally need adjoint equivalents of the routines `do_statevars_diags` and `diagnostics_fill_state`, etc, which appropriately handle the time steps.

> we should have the code tested in verification experiments as much as possible. I can see that there's something in global_ocean.cs32x15/input_ad, maybe it's worth it to add a second stream with 2D and RS/RL variables. I suggest: 'ADJetan ', 'ADJqnet ', 'ADJempmr', 'ADJtaux ','ADJtauy '. Anything else?

I think that is sufficient at the moment, but seaice adjoint variables are added, maybe some of these could be added to `global_ocean.cs32x15/input_ad.seaice`. Additionally, adjoint ptracers could be tested in `global_oce_biogeo_bling/`, which I could add while getting the ptracers dump to work with the new routines.

> for frequency(3) = 1.,, the model stalls with a Floating point exception:

No idea what that's about. Added it to the list to dig into

> We should definitely add seaice and exf fields. I can try to do that, assuming that it happens in seaice_ad_dump.F and exf_adjoint_snapshots_ad.F, right?

That would be awesome, thank you! 🙏  I believe those are the places, although it seems like the exf routine is commented out of `addummy_in_stepping`, so off the top of my head I'm not sure if or where it's called.

mjlosch commented:
@timothyas 
I added my seaice and elf code, mainly for you to have a look. Unfortunately, it only works when `useSEAICE = .FALSE.,`, I have no idea why, message(s):
```
(PID.TID 0000.0001) - WARNING -   diag.#   114 : ADJuvel  (#   1 ) in outp.Stream: adjDiag
(PID.TID 0000.0001) - WARNING -   has not been filled (ndiag=  0 )
(PID.TID 0000.0001) WARNING DIAGNOSTICS_OUT  => write ZEROS instead
```
etc. Tomorrow is another day.

mjlosch commented:
@timothyas 
now the only variable that's not filled ist ADJetan, because that's not implemented. It should be possible to do that better by calling the pair DIAGNOSTICS_SWITCH_ONOFF and DIAGNOSTICS_WRITE_ADJ explicitly at the beginning and at the end of forward step. Currently DIAGNOSTICS_SWITCH_ONOFF is called over and over again in TAF-generated code, which is probably OK, because useDiagnostics should be turned off, but still ...

Note, I am still not enirely sure, if the AD-diagnostics for exf and seaice are properly filled (I mean the correct number of times).

mjlosch commented:
@timothyas 
I am not sure if I understand the issue with diagnostic ADJetan. If we fill it in addummy_for_etan.F it may be half a time step away from the other variables, but if it is zero, when addummy_in_stepping.F is called, then that's even less useful. I'd rather have a slightly inconsistent diagnostic rather than none, or do I miss something important in the grander scheme of things?

Another question: does it make to to move calling DIAGNOSTICS_SWITCH_ONOFF to autodiff_inadmode_set.F and calling DIAGNOSTICS_WRITE_ADJ to autodiff_inadmode_unset.F? These routines are called at the beginning and end of the AD-forward step, and this would then catch all diagnostics being filled along the way. Again I may be missing something important here.

mjlosch commented:
commit 4434b0a implements the ideas of my previous comment. Diagnostics variable ADJetan is filled, but with an offset of 1 time step. If this is a problem, we can revert, but I think it is better than not having it at all, it should just be documented somewhere.

timothyas commented:
Hey @mjlosch, thanks for pushing this. I did not have a "master plan" for etan. To be honest, I avoided writing it out "improperly" in the earlier PR that I referenced basically because I was still new to everything, and couldn't really decide what to do so it's just good to hear your input. I'm fine with printing adj etan at a (reverse) lag, and we can just note that in the docs that this is the way it is.

As I mentioned in my email, I think moving the SWITCH_ONOFF and WRITE_ADJ calls to the SET/UNSET routines makes sense, and is generally cleaner. I don't see a reason not to do that, but will just look things over when I modify the ptracers.

timothyas commented:
@mjlosch 
- I created a branch on verification_other to run these diagnostics, let me know what you think: https://github.com/timothyas/verification_other/tree/test_adj_diags
- added ptracers to the mix, is there a place we can add this for a verification experiment?

timothyas commented:
Hey @mjlosch just a heads up I rebased this on top of the most recent master branch. It should not create any conflicts, but just letting you know before you pull next.

(this is one of a many branches making up my PhD work, and I had to rebase it to sync it up with the others and string them all together ... it's getting a little ridiculous)

mjlosch commented:
@timothyas 
> just a heads up I rebased this on top of the most recent master branch. It should not create any conflicts, but just letting you know before you pull next.

what does that mean? can I just pull or will I have to do something different?

timothyas commented:
Oh yep you can just pull, shouldn't be an issue

mjlosch commented:
@timothyas 
I don't know if I understand this rebase stuff. I had to resolve conflicts, which I would have had to commit before going, so I deleted my local branch and check it out again. Now there are only very few point on the list, right? I'll have a look at the frequency=1. error now. And then we can modify the experiments global_ocean.cs32x15.seaice and global_oce_biogeo_bling to get the code tested.

mjlosch commented:
@timothyas 
the frequency = 1. error has nothing to do with this PR. I also happens for pure forward simulations: I am testing with global_ocean.cs32x15.seaice, which has a startTime = 72000*86400 = 6.2208E+09
In diagnostics_out.F there's
https://github.com/MITgcm/MITgcm/blob/830e1a2c4c6f4f46908cffb59c28ba6c47110bc8/pkg/diagnostics/diagnostics_out.F#L126-L128
so when `freq=1`, the `INT(very large number)` stalls; for `freq=-1`, this code is not reached, and for `freq=86400.`, which would be appropriate (=deltaTclock in this experiment), everything works.
Question for @jm-c : is there a good way out? Can we "normalise" `timeRec(1)` like this `timeRec(1) = myTime-startTime- deltaTClock*0.5 _d 0`? Is this worth an issue?

timothyas commented:
Sorry for the troubles @mjlosch - I'll chill out with the rebasing... 🙃

Nice debugging @mjlosch. What about adding in `diagnostics_readparms` something like:

```
IF (frequency(l).GE. zeroRL) THEN
    freq(n) = MAX(frequency(l), deltaTclock)
ELSE
    freq(n) = MIN(frequency(l),-deltaTclock)
ENDIF
```

Thinking about that e.g. here 
https://github.com/MITgcm/MITgcm/blob/bd16afdcdefcbc67ac6c141b9a1d59a84f529014/pkg/diagnostics/diagnostics_readparms.F#L284-L289

But maybe more appropriate in a separate issue/pr.

---

I added the following:
- documentation regarding `ADJetan` and update list of available adjoint diagnostics
- add ADJ exf variables to global_ocean.cs32x15.seaice 
- add ADJ ptracer to global_oce_biogeo_bling

Let me know what you think, I'm happy to change any of those things around. Since that completes the original checklist I'll take this off draft mode, but the Work In Progress label is perhaps still appropriate...

mjlosch commented:
@timothyas 
great, thank you for the documentation part. Looks like we are done. Let's have someone else review this. Any suggestions? I'd always suggest @jm-c but I am sure he'd be happy if someone else can have a look at this too.

---
Let's not worry about the diagnostics frequency here. I like your suggestion. It will handle stupid user errors like mine, but it will not solve the actual problem of integer overflow if `timeRec(1)` becomes too large, I am thinking of long paleo simulations. I'll start an issue with this.

mjlosch commented:
@ifenty 
would you mind reviewing this PR?

ifenty commented:
@timothyas @mjlosch This is such a marvelous contribution.  I can see a lot of attention to detail here.  I'll take a quick look by generating new diag adj output from the global 3D and labsea verification experiments.

jrscott commented:
@mmazloff do you think you will have time to take a look at this/review in the next week or so?

mmazloff reviewed (approved):
I had a good look at code changes and all looks great! Very nice!
I'll try to find time to test it myself in the coming week, but since it passed the global_oce_biogeo_bling
verification test I am happy with it being checked in ASAP.

timothyas commented:
@mjlosch I have resolved the conflicts on my local branch while rebasing this to master, shall I push this?

mjlosch commented:
@timothyas go ahead. I'll try to update my local branch following your instructions as a test (o:

timothyas commented:
Great, done. @mjlosch you should be able to update your local branch with

```
git pull --rebase timothyas ptracers_adj_diags
```

or

```
git fetch timothyas
git rebase timothyas/ptracers_adj_diags
```

or replacing `timothyas` with whatever you have called my fork

jm-c commented:
@timothyas I would like to make few minor adjustments + pass some tests (e.g., where RS is real*4). Thinking it would be easier
to get master (now that PR #406 is merged) merge into this PR since both are dealing  with AUTODIFF things. Can I do that ?

timothyas commented:
Sure thing @jm-c

mjlosch commented:
@timothyas @jm-c 
Merging master into this branch leads to an extra TAF recomputation warning in `forward_step` (`myTime`,`myIter`), at least I didn't see that before. I think this is benign and TAF deals with it properly (I guess) by saving and restoring these two variables, but it would still be good to understand where this comes from. I think it has to do with the fact that we are now passing these  variables to `autodiff_inadmode_set/unset_ad.F`, but I have no clue why.

mjlosch commented:
Moving the call to `AUTODIFF_INADMODE_UNSET` at the beginning for `FORWARD_STEP` behind resetting the model iteration counter and model time fixes this. So I guess, that's already the accepted solution?

jm-c commented:
Hi both,
I am going to push some minot changes there (iclucing fixing this one) so you don't need to rush.
Cheers,
Jean-Michel

On Fri, Mar 12, 2021 at 01:01:10AM -0800, Martin Losch wrote:
> @timothyas @jm-c 
> Merging master into this branch leads to an extra TAF recomputation warning in `forward_step` (`myTime`,`myIter`), at least I didn't see that before. I think this is benign and TAF deals with it properly (I guess) by saving and restoring these two variables, but it would still be good to understand where this comes from. I think it has to do with the fact that we are now passing these  variables to `autodiff_inadmode_set/unset_ad.F`, but I have no clue why.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/411#issuecomment-797343419

jm-c commented:
@timothyas & @mjlosch I think with latest fixes all ADM experiments pass with gfortran & -devel (halfpipe_streamice was not compiling and all 4 global_ocean.90x40x15 were failing to run due to seg-fault reated to adBottomDragFld).
The next thing I am going to push is:
 1) minor argument change in the 4 dump_adj_xy*.F routines ; this fix the compilation when I check for argument consistency
    (with intel compiler & -devel) between S/R call and how argument are defined.
 2) in the simplest of the 4, i.e., dump_adj_xy.F I am proposing 2 improvements + better doc of argument list + simplification
  (i.e., removing un-needed included header files):
    a) avoid calling 2 times COPY_ADVAR_OUTP (for this purpose, I permuted the doDump and filling the diagnostics);
        this save one EXCH call (in this particular case); could think of going 1 step further (checking with DIAGNOSTICS_IS_ON)
        but not sure it's worth here, at this point.
     b) I dont' fill any diagnostics if diagName is '- None -' : This allows user to use this S/R to add a new AD-var output even 
        when a corresponding diagnostics does not exist (yet).
     c) in term of doc: I copied the description of "vType" from file: copy_advar_outp.F ; I think it's useful to have some comments
        but not sure if it's the right one.
      d) not checked yet with OpenAD (but need to since I remove some included header files)
 I suggest you review/improve the suggested changes in dump_adj_xy.F and then we could make similar changes in the 3 others.

timothyas commented on pkg/autodiff/dump_adj_xy.F:
```diff
@@ -117,45 +84,60 @@ SUBROUTINE DUMP_ADJ_XY( var2DRS, var2DRL, diagName, dumpName,
+      IF ( useDiag4AdjOutp .AND. diagName.NE.'- None -' ) THEN
+        IF ( dumpAdVarExch.EQ.2 ) THEN
+          IF ( .NOT.doDump )
+     &    CALL COPY_ADVAR_OUTP( var2DRS, var2DRL,
+     &                          var2Dc, 1, vType, myThid )
```
Good catch @jm-c, thanks for this. I would make the microscopic aesthetic suggestion of tabbing over the `CALL COPY_ADVAR_OUTP` lines to further distinguish from the `DIAGNOSTICS_FILL` right below.

timothyas commented:
Thanks for your changes @jm-c! I agree with all of them - especially a) and b) are nice additions.

In response to c) thanks for adding the `vType` description, that is the correct one since this is simply getting passed directly to that S/R.

As for OpenAD, it would be great if you could check this, I currently don't have a (quick) way to run it.

Lastly @mjlosch, I compiled and ran `global_oce_biogeo_bling` and it looks all good. It seems like fixing the placement of AUTODIFF_INADMODE_UNSET fixed the recomputations, since I'm not seeing anything unexpected in the log file. This seems like a clean solution

jm-c commented:
I can (and will) test OpenAD. It's not the most straightforward test to pass, so it take a little bit of time.

jm-c commented:
@timothyas And I forgot to mention regarding your suggestion for dump_adj_xy.F, if you don't mind pushing your improvement to this file, it would be great.

timothyas commented:
Not a problem - done. I'd be happy to copy the improvements you made on `_xy` to the other three similar files, just let me know if so.

jm-c commented:
I've checked (and fix) our full set of OpenAD experiments (including isomip by adding PR #442 in my test); I am going to push few minor cleaning, including removing exf_ad_dump.F file: I don't think there is any need to keep this old version which is not longer called since the end of 2007 and might confuse users who are looking at where exf-ad output are done.

jm-c reviewed (approved):
This is very useful addition for Ajoint run and is implemented nicely.
From my point of view, it is good to merge.
@timothyas if you want to check/comment on my latest changes, then after that we could go toward merging this PR.

timothyas commented:
Thanks so much for the cleanup and testing @jm-c, it all looks good to me. I just made a commit that deletes some of the includes in `addummy_for_etan` that you commented out. I can revert that if you disagree, but I thought I'd just push it since it's such a tiny change.

But yep, looks good to me - thanks everyone for the community effort on this one!

jm-c commented:
OK, going to merge this PR soon. And regarding tag-index addition, will add a little more that what is suggested here at the top of this PR, e.g., seaice & exf ad-diags.

ifenty commented:
@jm-c Can you please do this merge ASAP?
